Add application command guide (#4887)
hammerandchisel/discord-api-docs: docs/tutorials/Upgrading_to_Application_Commands.md

![Client interfaces showing the different types of application commands](command-types.png)

> info
> If you are developing an app for the first time, the [commands documentation](#DOCS_INTERACTIONS_APPLICATION_COMMANDS) may be a more helpful resource for you.

Before diving in, it’s worth noting that with the message content changes, apps can still access message content in their DMs with users, as well as when messages are sent that directly `@mentions` their bot user (since there is clear user intent).

And for apps that use message content in ways that can’t be ported to commands, you should review the [message content intent review policy](https://support-dev.discord.com/hc/en-us/articles/5324827539479) to see if you’re eligible to apply for the privileged intent.

---

## Types of commands

There are three types of application commands: slash commands, user commands, and message commands. When building your app, whether migrating or starting from scratch, consider which type of command(s) makes sense for a given feature of your app.

> ### Slash commands

Reviewer: ```suggestion
### Slash Commands
```

> [Slash commands](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/slash-commands) are the most common type of command. They are accessed by typing the command’s name prefixed with a backslash (`/`) or by using the plus button (+) to the left of the message input.

Reviewer 2: ```suggestion
[Slash commands](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/slash-commands) are the most common type of command. They are accessed by typing the command’s name prefixed with a forward slash (`/`) or by using the plus button (+) to the left of the message input.
```

Slash commands can appear in channels and DMs, so they’re helpful when an action is tied to a channel, a server, or nothing at all.

> ### User commands

Reviewer: ```suggestion
### User Commands
```

[User commands](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/user-commands) are in the context menu for users, which is accessed by right-clicking or tapping on any user. They’re helpful when an action is tied to an individual user, like a moderation app adding a timeout to someone.

> info
> Within the context menus for users and messages, the `Apps` submenu will only appear if an app in that server has installed a corresponding command (whether or not an individual user can use the installed command).

> ### Message commands

Reviewer: ```suggestion
### Message Commands
```

[Message commands](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/message-commands) are in the context menu for messages, which is accessed by right-clicking on a message, or by clicking on the ellipses at the top-right of a message. They’re helpful when an action is tied to a message, like creating a reminder to reply to a message the following day.

### After this section
- Start thinking about how different app features might map to the different types of application commands.

## Registering commands

> Commands can be registered via HTTP requests after an app is authorized with the `applications.commands` scope. Since commands aren’t tied to bot users, being authorized with the `bot` scope isn’t sufficient. Any commands created before an app is added to a server will automatically be installed.

Reviewer 3: ~~This implies that you need to send a request so that global commands will populate in a guild~~

Reviewer 3: (whoops, didn't read the whole paragraph, although it still implied that guild commands were global)

    "options": [
        {
            "name": "type",
            "description": "Type of rock",
            "type": 3,
            "required": True,
            "choices": [
                {
                    "name": "Igneous",
                    "value": "rock_igneous"
                },
                {
                    "name": "Sedimentary",
                    "value": "rock_sedimentary"
                },
                {
                    "name": "Metamorphic",
                    "value": "rock_metamorphic"
                }
            ]
        }
    ]
 }
```

#### Default permissions

When [creating](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/create-global-application-command) or [updating](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/edit-global-application-command) a command, you can use the `default_member_permissions` field to restrict the command to a set of bitwise permissions that reflect the default permission flags a user must be granted in order to use the command. 

In addition, the `dm_permission` flag can be used to control whether a global command is available in DMs (not available for guild commands).

#### Updating permissions

> Permissions for specific roles, users, and channels can be updated by your app if you have the `applications.commands.permissions.update` scope and a Bearer token that’s authenticated by a user with the necessary user permissions. Details about updating a command’s permissions are [in the documentation](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/permissions).

Reviewer 3: maybe make "Bearer token" a link to the OAuth2 docs?

### After this section
- Update your app to use the `applications.commands` scope
- Decide whether each of your commands should be scoped globally (available to all servers) or to a guild (available to specific servers)
- Decide what permissions your commands may need to run
- Create your app’s commands using HTTP endpoints

## Handling commands

> Commands use the [interactions model](#DOCS_INTERACTIONS_RECEIVING_AND_RESPONDING) rather than WebSocket-based gateways. The interactions model uses HTTP requests to send your app relevant information only when a Discord user triggers one of your app’s interactions.

Reviewer 3: ```suggestion
Commands use the [interactions model](#DOCS_INTERACTIONS_RECEIVING_AND_RESPONDING) rather than the WebSocket gateway. The interactions model uses HTTP requests to send your app relevant information only when a Discord user triggers one of your app’s interactions.
```

2. **Verifying request signatures**: To ensure that requests are coming from Discord, your endpoint must verify each request using the included headers, specifically `X-Signature-Ed25519` and `X-Signature-Timestamp`. If the signature fails validating, your app should return a `401` response. More information and example code can be found in the [Security and Authorization documentation](#DOCS_INTERACTIONS_RECEIVING_AND_RESPONDING/security-and-authorization).

> info
> Many libraries on the [Community Resources page](#DOCS_TOPICS_COMMUNITY_RESOURCES/interactions) simplify verification and interaction request handling by exporting reusable functions and/or handling it automatically.

After your URL is set up to handle signature verification and `PING` requests, you can add your Interaction Endpoint URL by navigating to your app settings from the [developer portal](https://discord.com/developers/applications). On the **General Information** page, you’ll see a field for your **Interactions Endpoint URL**. 

![Interactions endpoint URL in app settings](interactions-url.png)

After you paste your URL and click **Save Changes**, Discord will attempt to verify your endpoint.

> If all goes well, your endpoint will successfully save. And if not, you should double check that your app is running with a public endpoint, is verifying request signatures, and is properly responding to `PING` requests.

Reviewer 3: ```suggestion
If all goes well, your endpoint will be successfully saved. And if not, you should double check that your app is running with a public endpoint, is verifying request signatures, and is properly responding to `PING` requests.
```

As mentioned above, these include information relevant to handling the command like its name, the user who invoked it, and the guild and channel it was invoked from. It also includes additional details that could be relevant, like the [command options](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/application-command-object-application-command-option-structure) or [locale information](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/localization).

Since slash commands (`CHAT_INPUT` commands) are run in the context of a channel, you’ll notice that their payloads don’t contain any information about specific messages. If your app needs the message content, you can use [message commands](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/message-commands) which *do* include the message content.

> info
> In the getting started guide’s repository, there’s a code sample showing [how to create and handle commands using JavaScript](https://github.com/discord/discord-example-app/blob/main/examples/command.js).

Below is an example payload your app would receive when a user invoked a global command with an option:

> ```

Reviewer 3: examples throughout the documentation contain actual values resembling ones you see in practice, maybe they should be here too?

Author: I thought for this payload specifically, it might be helpful to notate the payload instead so it's clear which nested `id` fields are important to pay attention to, and how different fields map to features/concepts

As to where to communicate the changes, you can start with any communication channels you own—a support server, blog, or social media account for example.

You can also inform users about changes within the servers your app is installed in as long as it’s done in a non-intrusive way. If your app has a dedicated channel in posts, that would be a good place.

> warn
> Don’t DM all users in a server where your app is installed. It could get your app rate limited, but more importantly, it can be pretty intrusive and might lead to your app being uninstalled.

#### Example

The following is an example of an app update that may be sent to communicate the new way to access commands. Depending on the different features your app adopts (like options, subcommands, permissions, etc.), an update message or changelog entry will look very different.

![Example message updating users about a new feature](examples-update-message.png)

### Adding reactive education

In addition to proactively communicating changes, you can reactively educate users when making changes to your app.

> For example, if your app currently replies to messages that start with `!poll`, you can continue listening to those messages for a couple of weeks and redirect users to a new `/poll` command using an ephemeral message.

Reviewer 3: You can't respond to normal messages ephemerally

Author: rip sry will fix 🤦
